src/main/scala/intellij/haskell/external/component/PathComponent.scala

```diff
@@ -0,0 +1,119 @@
+package intellij.haskell.external.component
+
+import java.io.File
+import java.net.URL
+
+import com.intellij.openapi.project.Project
+import com.intellij.util.io.URLUtil
+import intellij.haskell.HaskellNotificationGroup
+import intellij.haskell.external.commandLine.StackCommandLine
+import org.apache.commons.io.FileUtils
+
+import scala.io.Source
+
+object PathComponent {
+  private final val PackageNamePattern = """.* (.*) [==|installed].*""".r
+  private final val LocalPkgdbResolverPattern = """.*/(lts-\d+\.\d+)/.*\n""".r
+  private final val CabalConfigFileResolverPattern = """.*/(lts-\d+\.\d+)""".r
+
+  def getIdeaPath(project: Project): String = {
+    project.getBasePath + File.separator + ".idea"
+  }
+
+  def getConfigFilePath(project: Project): String = {
+    getIdeaPath(project) + File.separator + "cabal.config"
+  }
+
+  def getAllAvailablePackageNames(project: Project): Option[Iterable[String]] = {
+    val configFilePath = getConfigFilePath(project)
+    val configFile = new File(configFilePath)
+    if (!configFile.exists()) {
+      downloadAndParseCabalConfigFile(project)
+    } else {
+      val needUpdate = for {
+        oldResolver <- getResolverFromCabalConfigFile(project)
+        newResolver <- getResolverFromLocalPkgdb(project)
+      } yield oldResolver != newResolver
+
+      needUpdate.flatMap(b => {
+        if (b) {
+          removeCabalConfig(project)
+          downloadAndParseCabalConfigFile(project)
+        } else {
+          parseCabalConfigFile(project)
+        }
+      })
+    }
+  }
+
+  private def downloadAndParseCabalConfigFile(project: Project): Option[Iterable[String]] = {
+    downloadCabalConfig(project).flatMap(b => {
+      if (b) {
+        parseCabalConfigFile(project)
+      } else {
+        Some(List())
+      }
+    })
+  }
+
+  private def parseCabalConfigFile(project: Project): Option[Iterable[String]] = {
+    val configFilePath = getConfigFilePath(project)
+    try {
+      Some(Source.fromFile(configFilePath).getLines().filter(!_.startsWith("--")).map {
+        case PackageNamePattern(packageName) => packageName
+        case _ => ""
+      }.filter(!_.isEmpty).toList)
+    } catch {
+      case _: Exception => None
+    }
+  }
+
+  def downloadCabalConfig(project: Project): Option[Boolean] = {
```

Comment:
I think we can keep the return type here since we should use the return result to check whether we successfully download the cabal.config file.